---
title: "EMO-BON Metagenomics: From Backend Integration to Frontend Processing"
date: 17 November 2025
author: David Paleƒçek [dpalecek@ualg.pt](mailto:dpalecek@ualg.pt)
editor: source
---

::: {.callout-note collapse="true" title="Last change"}
November 26, 2025
:::

EMO-BON establishes a long-term omics observatory for marine biodiversity through bimonthly sampling of coastal waters and both soft and hard sediments across more than 20 stations in Europe. Standard operating procedures are applied for sample collection, sequencing, and workflow analysis to ensure consistent taxonomic and functional annotation. FAIR principles are implemented using appropriate ontologies, RO-Crates, and a Python-based data analysis toolkit prepared for deployment in Virtual Research Environments (VREs).

## Session 1

Online session hosted online on 3rd November ([slides](https://github.com/Py-ualg/biohap/blob/main/assets/2511_dpalecek/DataStewards_emobon_david.pdf))

## Session 2

Faculty of Pharmacy of the University of Porto, 17th November during the ELIXIR Portugal All Hands 2025

This tutorial provides a set of minimal examples on how to generate a knowledge graph from set of RO-Crates, and use it as a sparQL endpoint either directly or using python `rdflib`. How to organize the data and build the RO-Crates is beyond scope, but [galaxy tutorial](https://training.galaxyproject.org/topics/fair/tutorials/ro-crate-in-python/tutorial.html) is a great starting point.

Second part focused on a specific use case of EMO-BON data, the pilot implementation of the Virtual Research Environment is introduced, basic visualization and analysis demonstrated, however, more extensive resources will be provided when the VRE is fully deployed and operational.

## Environmental setup

This setup serves is ubiquitus for both parts of the tutorial. We will clone two repositories and install one of them. To keep everything exactly the same, we create a designated folder

```bash
mkdir emobon_demo
cd emobon_demo
```

Create a dedicated conda/python environment

```bash
# if you are using conda
conda create -n "emobon" python=3.10   # or higher
conda activate emobon
```

### Python resources

Jupyter notebooks are included in the tutorials [repository](https://github.com/Py-ualg/biohap), which we clone locally

```bash
# create a folder
mkdir biohap
cd biohap

# clone the repository into newly created folder
git clone https://github.com/Py-ualg/biohap.git

# step into the repository
cd biohap

# install dependencies using pip
pip install -r requirements-biodata.txt

# setup jupyter kernel
ipython kernel install --user --name "biohap"
```

Separate setup is provided for the Data Analysis Toolkit section.

### Backend setup

As we are going to work with [EMO-BON](https://www.embrc.eu/emo-bon/) metagenomics data, one way is to download some of the RO-Crates [GitHub repository](https://github.com/emo-bon/analysis-results-cluster-01-crate) manually.

Easier, however. is to clone the whole repository. Navigate back to the `..../emobon_demo/` and do:

```bash
mkdir ro-crates
cd ro-crates

git clone https://github.com/emo-bon/analysis-results-cluster-01-crate.git
```

As you see in the GH as well, you have now the RO-Crates per sample and corresponding `.ttl` files to simplify your life.

We will use [fuseki](https://jena.apache.org/documentation/fuseki2/index.html) server for a SPARQL endpoint. It is a java ndpoint, therefore, you might need to install `java` first if `java -version` command does not return anything or your version is <17.0.

:::{.panel-tabset}

#### Linux

```bash
sudo apt update
sudo apt install -y openjdk-17-jre-headless
java -version
```

Here comes the fuseki download itself. Please chose appropriate folder for this. For direct download visit this [page](https://jena.apache.org/download/index.cgi).

```bash
wget https://dlcdn.apache.org/jena/binaries/apache-jena-fuseki-5.6.0.tar.gz

# open the archive
tar -xvf apache-jena-fuseki-5.6.0.tar.gz
```

#### Windows

For java, if you have `winget`, you can follow [these](https://learn.microsoft.com/en-us/java/openjdk/install#install-on-windows-with-the-windows-package-manager-winget) steps (not tested by the authors). Otherwise, download OpenJDK `.exe` file from [microsoft](https://learn.microsoft.com/en-us/java/openjdk/download) and install it. You might need to close and reopen the command line window to see the updated version with `java --version` command in your PowerShell.

Download directly the `fuseki` 5.6.0 zip file from [apache](https://jena.apache.org/download/index.cgi) and extract the archive.
:::

## Hands-on tutorial

:::{.panel-tabset}

### **Backend data integration**

{{< include 251117_part1.qmd >}}

### **Data Analysis Toolkit**

{{< include 251117_part2.qmd >}}

:::
