---
title: "RNA-Seq Data Analysis in R | From Counts to Biological Insights"
author: Isabel Duarte [giduarte@ualg.pt](mailto:giduarte@ualg.pt)
date: 04 june 2025
format:
  html:
    theme: cosmo
    toc: true
    toc-depth: 4
    toc-location: left
    code-fold: show
    code-summary: "Code"
    fig-width: 12
    fig-height: 6
    out.width: "100%"
    code-overflow: wrap
editor: source
---

## RNA-seq Analysis Crash Course

Welcome! This crash course walks you through RNA-seq data analysis in R: From raw count matrices to functional interpretation of differentially expressed genes.

#### Learning Objectives

1.  Perform quality control of the count data.
2.  Identify the R commands needed to run a differential expression analysis using `DESeq2`.
3.  Visualize the DE results.
4.  Perform functional enrichment of the DE genes.
5.  Visualize functionally enriched categories.

#### References

Attribution | These course materials are freely adapted from a combination between the following resources:

  - *nf-core RNA-seq* tutorial: [nf-core/rnaseq tutorial](https://nf-co.re/rnaseq/3.18.0/docs/usage/deanalysis/de_rstudio)
  - *The Carpentries Incubator* guided tutorial: [RNA-seq analysis with Bioconductor](https://carpentries-incubator.github.io/bioc-rnaseq/)
  - *The Harvard Chan Bioinformatics Core* tutorial: [Differential Gene Expression Analysis (bulk RNA-seq)](https://hbctraining.github.io/Intro-to-DGE/schedule/links-to-lessons.html)

Newer versions of this tutorial based on different dataset can be found [here](https://instructr.github.io/rnaseq_counts2bio/).

## Hands-on tutorial

:::{.panel-tabset}

### **Introduction**

{{< include 250604_iduarte_intro.qmd >}}


### **Tutorial**

{{< include 250604_iduarte_nfcore.qmd >}}

:::



