<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.22">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Example to query SparQL endpoint – BioHap: Bioinformatics, Handy and Practical</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-7b4406b7675125bc2ba204020e191172.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-42fb0fd9f3706fcb29599d1a10f6eca3.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="nav-sidebar docked nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../logo_ccmar.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">BioHap: Bioinformatics, Handy and Practical</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Example to query SparQL endpoint</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
        <div class="sidebar-tools-main">
    <a href="https://github.com/Py-ualg/biohap" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
</div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Introduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">BIOinformatics: Handy and Practical</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">BioData.pt</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../biodata_pt/intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Training on accessing, processing, and curating datasets 2025</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../biodata_pt/training_2511.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">EMO-BON Metagenomics: From Backend Integration to Frontend Processing</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Fall 2025</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../2025_fall/call.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Call for contributions and feedback</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Spring 2025</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../2025_spring/250416_atkacz.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Simple Linux – An Introduction to Genomic Analysis on Linux</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../2025_spring/250507_taires.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Microbiome Analysis of Amplicon Sequencing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../2025_spring/250604_iduarte/250604_iduarte.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">RNA-Seq Data Analysis in R | From Counts to Biological Insights</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Basics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../basics/abundance_tables.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Abundance table preprocessing</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#query-existing-fuseki" id="toc-query-existing-fuseki" class="nav-link active" data-scroll-target="#query-existing-fuseki">Query existing fuseki</a></li>
  <li><a href="#ssu-taxonomy-display" id="toc-ssu-taxonomy-display" class="nav-link" data-scroll-target="#ssu-taxonomy-display">SSU taxonomy display</a></li>
  <li><a href="#translate-ssu-triples-into-a-taxonomy-df" id="toc-translate-ssu-triples-into-a-taxonomy-df" class="nav-link" data-scroll-target="#translate-ssu-triples-into-a-taxonomy-df">Translate SSU triples into a taxonomy DF</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Example to query SparQL endpoint</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div id="035908fa" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> requests</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> json</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> re</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> rdflib <span class="im">import</span> Graph, Namespace, URIRef, RDF</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> urllib.parse <span class="im">import</span> quote_plus, urlparse</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> utils <span class="im">import</span> sparql_json_to_df</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>pd.options.display.max_columns <span class="op">=</span> <span class="va">None</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>pd.set_option(<span class="st">"display.max_colwidth"</span>, <span class="va">None</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="query-existing-fuseki" class="level2">
<h2 class="anchored" data-anchor-id="query-existing-fuseki">Query existing fuseki</h2>
<ul>
<li>assumed that it is populated with EMO-BON RO-Crates</li>
<li><code>fuseki</code> runs locally on the port <code>localhost:3030</code></li>
</ul>
<section id="how-many-triples-do-we-have" class="level4">
<h4 class="anchored" data-anchor-id="how-many-triples-do-we-have">How many triples do we have?</h4>
<div id="4d1d0bd1" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>q <span class="op">=</span> <span class="st">"""</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="st">SELECT (COUNT(*) AS ?c)</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="st">WHERE { </span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="st">    ?s ?p ?o</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="st">}</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>r <span class="op">=</span> requests.get(<span class="st">"http://localhost:3030/emobon"</span>, params<span class="op">=</span>{<span class="st">"query"</span>: q}, headers<span class="op">=</span>{<span class="st">"Accept"</span>: <span class="st">"application/sparql-results+json"</span>})</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(r.json())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'head': {'vars': ['c']}, 'results': {'bindings': [{'c': {'type': 'literal', 'datatype': 'http://www.w3.org/2001/XMLSchema#integer', 'value': '2637'}}]}}</code></pre>
</div>
</div>
<div id="473f8133" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> sparql_json_to_df(r.json())</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      c
0  2637</code></pre>
</div>
</div>
</section>
<section id="filter-our-all-the-texthtml-files" class="level4">
<h4 class="anchored" data-anchor-id="filter-our-all-the-texthtml-files">Filter our all the <code>text/html</code> files</h4>
<div id="c273f583" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>q <span class="op">=</span> <span class="st">"""</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="st">PREFIX sdo: &lt;http://schema.org/&gt;</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="st">SELECT ?x ?dtype</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="st">WHERE {</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="st">  ?x sdo:encodingFormat ?dtype .</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="st">  FILTER regex(str(?dtype), "^text/html", "i")</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="st">}</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>r <span class="op">=</span> requests.get(<span class="st">"http://localhost:3030/emobon"</span>, params<span class="op">=</span>{<span class="st">"query"</span>: q}, headers<span class="op">=</span>{<span class="st">"Accept"</span>: <span class="st">"application/sparql-results+json"</span>})</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> sparql_json_to_df(r.json())</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">x</th>
<th data-quarto-table-cell-role="th">dtype</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>https://www.ebi.ac.uk/ena/browser/view/ERS20568987</td>
<td>text/html</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/krona.html</td>
<td>text/html</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/krona.html</td>
<td>text/html</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>https://www.ebi.ac.uk/ena/browser/view/ERS20569004</td>
<td>text/html</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>https://www.ebi.ac.uk/ena/browser/view/ERS20569003</td>
<td>text/html</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_AAOT_Wa_67-ro-crate/fastp.html</td>
<td>text/html</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_HCMR-1_Wa_6-ro-crate/fastp.html</td>
<td>text/html</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/krona.html</td>
<td>text/html</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_OSD74_Wa_21-ro-crate/fastp.html</td>
<td>text/html</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_AAOT_Wa_66-ro-crate/fastp.html</td>
<td>text/html</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/krona.html</td>
<td>text/html</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">11</td>
<td>https://www.ebi.ac.uk/ena/browser/view/ERS20569012</td>
<td>text/html</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="return-also-the-sdodownloadurl-of-those-files." class="level4">
<h4 class="anchored" data-anchor-id="return-also-the-sdodownloadurl-of-those-files.">Return also the <code>sdo:downloadUrl</code> of those files.</h4>
<div id="f7c8f6f3" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>q <span class="op">=</span> <span class="st">"""</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="st">PREFIX sdo: &lt;http://schema.org/&gt;</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="st">SELECT ?x ?dtype ?durl</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="st">WHERE {</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="st">  ?x sdo:encodingFormat ?dtype ;</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="st">     sdo:downloadUrl ?durl .</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="st">  FILTER regex(str(?dtype), "^text/html", "i")</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="st">}</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>r <span class="op">=</span> requests.get(<span class="st">"http://localhost:3030/emobon"</span>, params<span class="op">=</span>{<span class="st">"query"</span>: q}, headers<span class="op">=</span>{<span class="st">"Accept"</span>: <span class="st">"application/sparql-results+json"</span>})</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> sparql_json_to_df(r.json())</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">x</th>
<th data-quarto-table-cell-role="th">dtype</th>
<th data-quarto-table-cell-role="th">durl</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>https://www.ebi.ac.uk/ena/browser/view/ERS20568987</td>
<td>text/html</td>
<td>https://www.ebi.ac.uk/ena/browser/view/ERS20568987</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/krona.html</td>
<td>text/html</td>
<td>https://s3.mesocentre.uca.fr/mgf-data-products/files/md5/1f/d7f1e97dc438433527d667ad7694da</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/krona.html</td>
<td>text/html</td>
<td>https://s3.mesocentre.uca.fr/mgf-data-products/files/md5/65/40f5e154fe4b60253aa2bb546aa163</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>https://www.ebi.ac.uk/ena/browser/view/ERS20569004</td>
<td>text/html</td>
<td>https://www.ebi.ac.uk/ena/browser/view/ERS20569004</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>https://www.ebi.ac.uk/ena/browser/view/ERS20569003</td>
<td>text/html</td>
<td>https://www.ebi.ac.uk/ena/browser/view/ERS20569003</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_AAOT_Wa_67-ro-crate/fastp.html</td>
<td>text/html</td>
<td>https://s3.mesocentre.uca.fr/mgf-data-products/files/md5/a3/01657724b6a1dd94b2d5c773ab3389</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_HCMR-1_Wa_6-ro-crate/fastp.html</td>
<td>text/html</td>
<td>https://s3.mesocentre.uca.fr/mgf-data-products/files/md5/12/e78d197511d96354e0996e9d674d39</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/krona.html</td>
<td>text/html</td>
<td>https://s3.mesocentre.uca.fr/mgf-data-products/files/md5/94/cf7ab813e63199fdfa4a134ee87884</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_OSD74_Wa_21-ro-crate/fastp.html</td>
<td>text/html</td>
<td>https://s3.mesocentre.uca.fr/mgf-data-products/files/md5/0f/28859f70d366611ab6c31110f865bb</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_AAOT_Wa_66-ro-crate/fastp.html</td>
<td>text/html</td>
<td>https://s3.mesocentre.uca.fr/mgf-data-products/files/md5/14/151c6e5052ad5fe322b710fcd83759</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/krona.html</td>
<td>text/html</td>
<td>https://s3.mesocentre.uca.fr/mgf-data-products/files/md5/5e/e412ca17b94731cdc691ec3c53efda</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">11</td>
<td>https://www.ebi.ac.uk/ena/browser/view/ERS20569012</td>
<td>text/html</td>
<td>https://www.ebi.ac.uk/ena/browser/view/ERS20569012</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="return-ssu-taxonomy-download-links" class="level4">
<h4 class="anchored" data-anchor-id="return-ssu-taxonomy-download-links">Return SSU taxonomy download links</h4>
<div id="8c9f8b33" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>q <span class="op">=</span> <span class="st">"""</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="st">PREFIX sdo: &lt;http://schema.org/&gt;</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="st">SELECT ?subject ?predicate ?object ?durl</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="st">WHERE {</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="st">  ?subject ?predicate ?object .</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="st">  FILTER regex(str(?object), "SSU-taxonomy-summary", "i")</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="st">  OPTIONAL { ?object sdo:downloadUrl ?durl }</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="st">}</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="st">LIMIT 50</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>r <span class="op">=</span> requests.get(</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"http://localhost:3030/emobon"</span>,</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>    params<span class="op">=</span>{<span class="st">"query"</span>: q},</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>    headers<span class="op">=</span>{<span class="st">"Accept"</span>: <span class="st">"application/sparql-results+json"</span>},</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> sparql_json_to_df(r.json())</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">subject</th>
<th data-quarto-table-cell-role="th">predicate</th>
<th data-quarto-table-cell-role="th">object</th>
<th data-quarto-table-cell-role="th">durl</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/</td>
<td>http://schema.org/hasPart</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl</td>
<td>https://s3.mesocentre.uca.fr/mgf-data-products/files/md5/f1/6ea0374e1b699258f80cb72d183c9d</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/</td>
<td>http://schema.org/hasPart</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl</td>
<td>https://s3.mesocentre.uca.fr/mgf-data-products/files/md5/89/eb7c8afaadc7a3b7e61181ad289ff3</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/</td>
<td>http://schema.org/hasPart</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl</td>
<td>https://s3.mesocentre.uca.fr/mgf-data-products/files/md5/06/c40e558f4e38f74d6010db8a68d9ca</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/</td>
<td>http://schema.org/hasPart</td>
<td>https://github.com/emo-bon/analysis-results-cluster-01-crate/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl</td>
<td>https://s3.mesocentre.uca.fr/mgf-data-products/files/md5/10/daf150ddf1ffdcd0f8e06bca453987</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
</section>
<section id="ssu-taxonomy-display" class="level2">
<h2 class="anchored" data-anchor-id="ssu-taxonomy-display">SSU taxonomy display</h2>
<ul>
<li>from the object values we see the taxonomy is in ttl format, which means it has been triplicated during so called <code>semantic uplift</code></li>
</ul>
<div id="47dc3145" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>url <span class="op">=</span> df[<span class="st">"durl"</span>].dropna().unique()[<span class="dv">0</span>]</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>r <span class="op">=</span> requests.get(url)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># save to a file</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">"ssu_example.ttl"</span>, <span class="st">"wb"</span>) <span class="im">as</span> f:</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    f.write(r.content)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="10ffe2c0" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>url</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="10">
<pre><code>'https://s3.mesocentre.uca.fr/mgf-data-products/files/md5/f1/6ea0374e1b699258f80cb72d183c9d'</code></pre>
</div>
</div>
<div id="2f22c789" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> Graph()</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>g.parse(<span class="st">"ssu_example.ttl"</span>, <span class="bu">format</span><span class="op">=</span><span class="st">"turtle"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert all triples into a list of tuples</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>triples <span class="op">=</span> [(<span class="bu">str</span>(s), <span class="bu">str</span>(p), <span class="bu">str</span>(o)) <span class="cf">for</span> s, p, o <span class="kw">in</span> g]</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a pandas DataFrame</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame(triples, columns<span class="op">=</span>[<span class="st">"subject"</span>, <span class="st">"predicate"</span>, <span class="st">"object"</span>])</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>df.head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>file:///home/david-palecek/coding/biohap/biohap/biohap/biodata_pt/NBs/?xml version="1.0" encoding="UTF-8"? does not look like a valid URI, trying to serialize this will break.</code></pre>
</div>
<div class="cell-output cell-output-error">
<div class="ansi-escaped-output">
<pre>Traceback <span class="ansi-cyan-fg">(most recent call last)</span>:

  File <span class="ansi-bright-green-fg">~/miniforge3/envs/biohap/lib/python3.12/site-packages/IPython/core/interactiveshell.py:3699</span> in <span class="ansi-bright-magenta-fg">run_code</span>
    exec(code_obj, self.user_global_ns, self.user_ns)

  Cell <span class="ansi-bright-green-fg">In[9]</span><span class="ansi-bright-green-fg">, line 2</span>
    g.parse("ssu_example.ttl", format="turtle")

  File <span class="ansi-bright-green-fg">~/miniforge3/envs/biohap/lib/python3.12/site-packages/rdflib/graph.py:1656</span> in <span class="ansi-bright-magenta-fg">parse</span>
    raise se

  File <span class="ansi-bright-green-fg">~/miniforge3/envs/biohap/lib/python3.12/site-packages/rdflib/graph.py:1647</span> in <span class="ansi-bright-magenta-fg">parse</span>
    parser.parse(source, self, **args)

  File <span class="ansi-bright-green-fg">~/miniforge3/envs/biohap/lib/python3.12/site-packages/rdflib/plugins/parsers/notation3.py:2034</span> in <span class="ansi-bright-magenta-fg">parse</span>
    p.loadStream(stream)

  File <span class="ansi-bright-green-fg">~/miniforge3/envs/biohap/lib/python3.12/site-packages/rdflib/plugins/parsers/notation3.py:492</span> in <span class="ansi-bright-magenta-fg">loadStream</span>
    return self.loadBuf(stream.read())  # Not ideal

  File <span class="ansi-bright-green-fg">~/miniforge3/envs/biohap/lib/python3.12/site-packages/rdflib/plugins/parsers/notation3.py:498</span> in <span class="ansi-bright-magenta-fg">loadBuf</span>
    self.feed(buf)

  File <span class="ansi-bright-green-fg">~/miniforge3/envs/biohap/lib/python3.12/site-packages/rdflib/plugins/parsers/notation3.py:524</span> in <span class="ansi-bright-magenta-fg">feed</span>
    i = self.directiveOrStatement(s, j)

  File <span class="ansi-bright-green-fg">~/miniforge3/envs/biohap/lib/python3.12/site-packages/rdflib/plugins/parsers/notation3.py:545</span> in <span class="ansi-bright-magenta-fg">directiveOrStatement</span>
    return self.checkDot(argstr, j)

  File <span class="ansi-bright-green-fg">~/miniforge3/envs/biohap/lib/python3.12/site-packages/rdflib/plugins/parsers/notation3.py:1227</span> in <span class="ansi-bright-magenta-fg">checkDot</span>
    self.BadSyntax(argstr, j, "expected '.' or '}' or ']' at end of statement")

<span class="ansi-cyan-fg">  </span><span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">~/miniforge3/envs/biohap/lib/python3.12/site-packages/rdflib/plugins/parsers/notation3.py:1743</span><span class="ansi-cyan-fg"> in </span><span class="ansi-magenta-fg">BadSyntax</span>
<span class="ansi-red-fg">    </span><span class="ansi-red-fg">raise BadSyntax(self._thisDoc, self.lines, argstr, i, msg)</span>

  <span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">&lt;string&gt;</span>
<span class="ansi-red-fg">    </span>
<span class="ansi-red-fg">BadSyntax</span>
</pre>
</div>
</div>
</div>
</section>
<section id="translate-ssu-triples-into-a-taxonomy-df" class="level2">
<h2 class="anchored" data-anchor-id="translate-ssu-triples-into-a-taxonomy-df">Translate SSU triples into a taxonomy DF</h2>
<div id="6fd1b9fd" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>COLUMNS <span class="op">=</span> [</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">'ncbi_tax_id'</span>, <span class="st">'abundance'</span>, <span class="st">'superkingdom'</span>,</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">'kingdom'</span>, <span class="st">'phylum'</span>, <span class="st">'class'</span>, <span class="st">'order'</span>,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">'family'</span>, <span class="st">'genus'</span>, <span class="st">'species'</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>RANKS <span class="op">=</span> [<span class="st">'superkingdom'</span>,<span class="st">'kingdom'</span>,<span class="st">'phylum'</span>,<span class="st">'class'</span>,<span class="st">'order'</span>,<span class="st">'family'</span>,<span class="st">'genus'</span>,<span class="st">'species'</span>]</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co"># ----- helper functions -----------------------------------------------------</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> as_str(node):</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="va">None</span> <span class="cf">if</span> node <span class="kw">is</span> <span class="va">None</span> <span class="cf">else</span> <span class="bu">str</span>(node)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> extract_ncbi_id(uri_str):</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Extract numeric taxid from NCBI URI or fragment like #41873."""</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> uri_str:</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="va">None</span></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>    m <span class="op">=</span> re.search(<span class="vs">r'</span><span class="pp">[?&amp;]</span><span class="vs">id=</span><span class="kw">(</span><span class="dv">\d</span><span class="op">+</span><span class="kw">)</span><span class="vs">'</span>, uri_str)</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> m:</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> m.group(<span class="dv">1</span>)</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>    m <span class="op">=</span> re.search(<span class="vs">r'#</span><span class="kw">(</span><span class="dv">\d</span><span class="op">+</span><span class="kw">)</span><span class="dv">$</span><span class="vs">'</span>, uri_str)</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> m:</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> m.group(<span class="dv">1</span>)</span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>    m <span class="op">=</span> re.search(<span class="vs">r'/</span><span class="kw">(</span><span class="dv">\d</span><span class="op">+</span><span class="kw">)</span>(?:<span class="dv">$</span><span class="cf">|</span><span class="pp">[/?#]</span>)<span class="vs">'</span>, uri_str)</span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> m:</span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> m.group(<span class="dv">1</span>)</span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="va">None</span></span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> last_path_segment(uri_str):</span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> uri_str:</span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="va">None</span></span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true" tabindex="-1"></a>    p <span class="op">=</span> urlparse(uri_str)</span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> p.fragment:</span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> p.fragment</span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true" tabindex="-1"></a>    seg <span class="op">=</span> p.path.rstrip(<span class="st">'/'</span>).split(<span class="st">'/'</span>)[<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> seg <span class="cf">if</span> seg <span class="op">!=</span> <span class="st">''</span> <span class="cf">else</span> <span class="va">None</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="864ff23a" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ----- load graph -----------------------------------------------------------</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> Graph()</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>g.parse(<span class="st">"ssu_example.ttl"</span>, <span class="bu">format</span><span class="op">=</span><span class="st">"turtle"</span>)   <span class="co"># &lt;-- change path if needed</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ----- namespaces used in your sample --------------------------------------</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>PROD <span class="op">=</span> Namespace(<span class="st">"https://data.emobon.embrc.eu/ns/product#"</span>)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>DCT  <span class="op">=</span> Namespace(<span class="st">"http://purl.org/dc/terms/"</span>)</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>SCHEMA <span class="op">=</span> Namespace(<span class="st">"http://schema.org/"</span>)  <span class="co"># not required here but safe</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="co"># ----- accumulate rows keyed by (source_material_ID, ncbi_tax_id) -----------</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>acc <span class="op">=</span> {}  <span class="co"># (source_id, taxid) -&gt; dict of columns</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> ensure_row(source_id, taxid):</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>    key <span class="op">=</span> (source_id <span class="kw">or</span> <span class="st">"unknown_sample"</span>, taxid <span class="kw">or</span> <span class="st">"unknown_taxid"</span>)</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> key <span class="kw">not</span> <span class="kw">in</span> acc:</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>        acc[key] <span class="op">=</span> {col: <span class="va">None</span> <span class="cf">for</span> col <span class="kw">in</span> COLUMNS}</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>        acc[key][<span class="st">'ncbi_tax_id'</span>] <span class="op">=</span> taxid</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> acc[key]</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a><span class="co"># ----- iterate TaxonomicAnnotation nodes -----------------------------------</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> ta_node <span class="kw">in</span> g.subjects(RDF.<span class="bu">type</span>, PROD.TaxonomicAnnotation):</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>    <span class="co"># read annotation-level properties</span></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>    sample_uri <span class="op">=</span> g.value(ta_node, PROD.ofSample) <span class="kw">or</span> g.value(ta_node, DCT.ofSample)</span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>    source_id <span class="op">=</span> last_path_segment(as_str(sample_uri))</span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>    identifier <span class="op">=</span> g.value(ta_node, DCT.identifier)  <span class="co"># expected to be NCBI URI</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>    identifier_s <span class="op">=</span> as_str(identifier)</span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>    ncbi_id <span class="op">=</span> extract_ncbi_id(identifier_s)</span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a>    <span class="co"># abundance (prod:ssuRNA) and otuID</span></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a>    abundance_term <span class="op">=</span> g.value(ta_node, PROD.ssuRNA)</span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a>    <span class="co"># sometimes abundance literal typed - convert to int if possible</span></span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a>    <span class="cf">try</span>:</span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a>        abundance <span class="op">=</span> <span class="bu">int</span>(<span class="bu">str</span>(abundance_term))</span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a>    <span class="cf">except</span> <span class="pp">Exception</span>:</span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a>        abundance <span class="op">=</span> <span class="bu">str</span>(abundance_term)</span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a>    <span class="co"># create/ensure row</span></span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a>    row <span class="op">=</span> ensure_row(source_id, ncbi_id)</span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a>    row[<span class="st">'abundance'</span>] <span class="op">=</span> abundance</span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a>    <span class="co"># now look up the taxon node (identifier) and extract rank / scientificName</span></span>
<span id="cb15-42"><a href="#cb15-42" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> identifier <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span>:</span>
<span id="cb15-43"><a href="#cb15-43" aria-hidden="true" tabindex="-1"></a>        tax_subject <span class="op">=</span> URIRef(identifier_s)</span>
<span id="cb15-44"><a href="#cb15-44" aria-hidden="true" tabindex="-1"></a>        sci_name_term <span class="op">=</span> g.value(tax_subject, DCT.scientificName) <span class="kw">or</span> g.value(tax_subject, DCT.title)</span>
<span id="cb15-45"><a href="#cb15-45" aria-hidden="true" tabindex="-1"></a>        rank_term <span class="op">=</span> g.value(tax_subject, DCT.taxonRank)</span>
<span id="cb15-46"><a href="#cb15-46" aria-hidden="true" tabindex="-1"></a>        sci_name <span class="op">=</span> as_str(sci_name_term)</span>
<span id="cb15-47"><a href="#cb15-47" aria-hidden="true" tabindex="-1"></a>        rank <span class="op">=</span> as_str(rank_term).lower() <span class="cf">if</span> rank_term <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span> <span class="cf">else</span> <span class="va">None</span></span>
<span id="cb15-48"><a href="#cb15-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-49"><a href="#cb15-49" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> rank <span class="kw">and</span> sci_name:</span>
<span id="cb15-50"><a href="#cb15-50" aria-hidden="true" tabindex="-1"></a>            <span class="co"># if rank is one of our expected ranks, store at that column</span></span>
<span id="cb15-51"><a href="#cb15-51" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> rank <span class="kw">in</span> RANKS:</span>
<span id="cb15-52"><a href="#cb15-52" aria-hidden="true" tabindex="-1"></a>                row[rank] <span class="op">=</span> sci_name</span>
<span id="cb15-53"><a href="#cb15-53" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb15-54"><a href="#cb15-54" aria-hidden="true" tabindex="-1"></a>            <span class="cf">raise</span> <span class="pp">ValueError</span>(<span class="ss">f"Missing rank or scientific name for taxon </span><span class="sc">{</span>identifier_s<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb15-55"><a href="#cb15-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-56"><a href="#cb15-56" aria-hidden="true" tabindex="-1"></a><span class="co"># ----- build final DataFrame ------------------------------------------------</span></span>
<span id="cb15-57"><a href="#cb15-57" aria-hidden="true" tabindex="-1"></a>out_rows <span class="op">=</span> []</span>
<span id="cb15-58"><a href="#cb15-58" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (source_id, taxid), vals <span class="kw">in</span> acc.items():</span>
<span id="cb15-59"><a href="#cb15-59" aria-hidden="true" tabindex="-1"></a>    <span class="co"># enforce ncbi_tax_id numeric when possible</span></span>
<span id="cb15-60"><a href="#cb15-60" aria-hidden="true" tabindex="-1"></a>    <span class="cf">try</span>:</span>
<span id="cb15-61"><a href="#cb15-61" aria-hidden="true" tabindex="-1"></a>        vals[<span class="st">'ncbi_tax_id'</span>] <span class="op">=</span> <span class="bu">int</span>(taxid) <span class="cf">if</span> (taxid <span class="kw">and</span> taxid <span class="op">!=</span> <span class="st">"unknown_taxid"</span>) <span class="cf">else</span> <span class="va">None</span></span>
<span id="cb15-62"><a href="#cb15-62" aria-hidden="true" tabindex="-1"></a>    <span class="cf">except</span> <span class="pp">Exception</span>:</span>
<span id="cb15-63"><a href="#cb15-63" aria-hidden="true" tabindex="-1"></a>        vals[<span class="st">'ncbi_tax_id'</span>] <span class="op">=</span> taxid</span>
<span id="cb15-64"><a href="#cb15-64" aria-hidden="true" tabindex="-1"></a>    vals[<span class="st">'source_material_ID'</span>] <span class="op">=</span> source_id</span>
<span id="cb15-65"><a href="#cb15-65" aria-hidden="true" tabindex="-1"></a>    out_rows.append(vals)</span>
<span id="cb15-66"><a href="#cb15-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-67"><a href="#cb15-67" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame(out_rows)</span>
<span id="cb15-68"><a href="#cb15-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-69"><a href="#cb15-69" aria-hidden="true" tabindex="-1"></a><span class="co"># set index like your example and reorder columns</span></span>
<span id="cb15-70"><a href="#cb15-70" aria-hidden="true" tabindex="-1"></a>df[<span class="st">'source_material_ID'</span>] <span class="op">=</span> df[<span class="st">'source_material_ID'</span>].fillna(<span class="st">'unknown_sample'</span>)</span>
<span id="cb15-71"><a href="#cb15-71" aria-hidden="true" tabindex="-1"></a>df[<span class="st">'ncbi_tax_id'</span>] <span class="op">=</span> df[<span class="st">'ncbi_tax_id'</span>].fillna(<span class="st">'unknown_taxid'</span>)</span>
<span id="cb15-72"><a href="#cb15-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-73"><a href="#cb15-73" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df.set_index([<span class="st">'source_material_ID'</span>, <span class="st">'ncbi_tax_id'</span>])</span>
<span id="cb15-74"><a href="#cb15-74" aria-hidden="true" tabindex="-1"></a>final_cols <span class="op">=</span> [<span class="st">'abundance'</span>] <span class="op">+</span> RANKS</span>
<span id="cb15-75"><a href="#cb15-75" aria-hidden="true" tabindex="-1"></a><span class="co"># ensure columns exist in the DataFrame before slicing</span></span>
<span id="cb15-76"><a href="#cb15-76" aria-hidden="true" tabindex="-1"></a>final_cols <span class="op">=</span> [c <span class="cf">for</span> c <span class="kw">in</span> final_cols <span class="cf">if</span> c <span class="kw">in</span> df.columns]</span>
<span id="cb15-77"><a href="#cb15-77" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df[final_cols].sort_index()</span>
<span id="cb15-78"><a href="#cb15-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-79"><a href="#cb15-79" aria-hidden="true" tabindex="-1"></a><span class="co"># convert abundance to numeric (if any)</span></span>
<span id="cb15-80"><a href="#cb15-80" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">'abundance'</span> <span class="kw">in</span> df.columns:</span>
<span id="cb15-81"><a href="#cb15-81" aria-hidden="true" tabindex="-1"></a>    df[<span class="st">'abundance'</span>] <span class="op">=</span> pd.to_numeric(df[<span class="st">'abundance'</span>], errors<span class="op">=</span><span class="st">'coerce'</span>)</span>
<span id="cb15-82"><a href="#cb15-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-83"><a href="#cb15-83" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df.shape)</span>
<span id="cb15-84"><a href="#cb15-84" aria-hidden="true" tabindex="-1"></a>df.head(<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(388, 9)</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="13">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">abundance</th>
<th data-quarto-table-cell-role="th">superkingdom</th>
<th data-quarto-table-cell-role="th">kingdom</th>
<th data-quarto-table-cell-role="th">phylum</th>
<th data-quarto-table-cell-role="th">class</th>
<th data-quarto-table-cell-role="th">order</th>
<th data-quarto-table-cell-role="th">family</th>
<th data-quarto-table-cell-role="th">genus</th>
<th data-quarto-table-cell-role="th">species</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">source_material_ID</th>
<th data-quarto-table-cell-role="th">ncbi_tax_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td rowspan="20" data-quarto-table-cell-role="th" data-valign="top">EMOBON_AAOT_Wa_66</td>
<td data-quarto-table-cell-role="th">2</td>
<td>783.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">29</td>
<td>5.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Myxococcales</td>
<td>None</td>
<td>None</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">81</td>
<td>2.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Hyphomicrobium</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">112</td>
<td>11.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Planctomycetales</td>
<td>None</td>
<td>None</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">136</td>
<td>1.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Spirochaetales</td>
<td>None</td>
<td>None</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">137</td>
<td>1.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Spirochaetaceae</td>
<td>None</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">226</td>
<td>15.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Alteromonas</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">237</td>
<td>12.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Flavobacterium</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">265</td>
<td>10.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Paracoccus</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">279</td>
<td>1.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Xanthobacter</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">286</td>
<td>10.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Pseudomonas</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">356</td>
<td>41.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Rhizobiales</td>
<td>None</td>
<td>None</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">374</td>
<td>35.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Bradyrhizobium</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">407</td>
<td>31.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Methylobacterium</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">444</td>
<td>13.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Legionellaceae</td>
<td>None</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">445</td>
<td>1.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Legionella</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">468</td>
<td>1.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Moraxellaceae</td>
<td>None</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">469</td>
<td>14.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Acinetobacter</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">482</td>
<td>2.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Neisseria</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">543</td>
<td>1.0</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Enterobacteriaceae</td>
<td>None</td>
<td>None</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>